crates/fuel-core/src/service/genesis.rs

```diff
@@ -141,160 +198,180 @@ fn import_genesis_block(
-                .insert(&contract_id, &(salt, root))?
-                .is_some()
-            {
-                return Err(anyhow!("Contract info should not exist"))
-            }
-            if db
-                .storage::<ContractsLatestUtxo>()
-                .insert(
-                    &contract_id,
-                    &ContractUtxoInfo {
-                        utxo_id,
-                        tx_pointer,
-                    },
-                )?
-                .is_some()
-            {
-                return Err(anyhow!("Contract utxo should not exist"))
-            }
-            init_contract_state(db, &contract_id, contract_config)?;
-            init_contract_balance(db, &contract_id, contract_config)?;
-            contracts_tree
-                .push(ContractRef::new(&mut *db, contract_id).root()?.as_slice());
-        }
+    let contract = Contract::from(contract_config.code.as_slice());
+    let salt = contract_config.salt;
+    let root = contract.root();
+    let contract_id = contract_config.contract_id;
+    let utxo_id = if let (Some(tx_id), Some(output_idx)) =
+        (contract_config.tx_id, contract_config.output_index)
+    {
+        UtxoId::new(tx_id, output_idx)
+    } else {
+        UtxoId::new(
+            // generated transaction id([0..[out_index/255]])
+            Bytes32::try_from(
+                (0..(Bytes32::LEN - WORD_SIZE))
+                    .map(|_| 0u8)
+                    .chain((output_index as u64 / 255).to_be_bytes().into_iter())
+                    .collect_vec()
+                    .as_slice(),
+            )
+            .expect("Incorrect genesis transaction id byte length"),
+            output_index as u8,
+        )
+    };
+    let tx_pointer = if let (Some(block_height), Some(tx_idx)) = (
+        contract_config.tx_pointer_block_height,
+        contract_config.tx_pointer_tx_idx,
+    ) {
+        TxPointer::new(block_height, tx_idx)
+    } else {
+        TxPointer::default()
+    };
+
+    if tx_pointer.block_height() > height {
+        return Err(anyhow!(
+            "contract tx_pointer cannot be greater than genesis block"
+        ));
+    }
+
+    // insert contract code
+    if db
+        .storage::<ContractsRawCode>()
+        .insert(&contract_id, contract.as_ref())?
+        .is_some()
+    {
+        return Err(anyhow!("Contract code should not exist"));
+    }
+
+    // insert contract root
+    if db
+        .storage::<ContractsInfo>()
+        .insert(&contract_id, &(salt, root))?
+        .is_some()
+    {
+        return Err(anyhow!("Contract info should not exist"));
+    }
+    if db
+        .storage::<ContractsLatestUtxo>()
+        .insert(
+            &contract_id,
+            &ContractUtxoInfo {
+                utxo_id,
+                tx_pointer,
+            },
+        )?
+        .is_some()
+    {
+        return Err(anyhow!("Contract utxo should not exist"));
     }
-    Ok(contracts_tree.root())
+    Ok(())
```

Comment:
Do you plan to return hash of the whole contract?

Reply:
See https://github.com/FuelLabs/fuel-core/pull/1474#discussion_r1463488334